### PR TITLE
PAYARA-2263 per-jdk version domain options

### PR DESCRIPTION
**Pull request by lprimak**
Ability to enable / disable JDK options based on JDK version pattern

### CODE REVIEW

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/commands/DeleteJvmOptions.java [hunk 1]
```diff
@@ -159,20 +159,18 @@ public Object run(JavaConfig jc) throws PropertyVetoException, TransactionFailur
...
-                }
-                return true;
+        SingleConfigCode<JvmOptionBag> scc = (JvmOptionBag bag1) -> {
+            List<String> jvmopts = new ArrayList<>(bag1.getJvmRawOptions());
+            int orig = jvmopts.size();
+            boolean removed = jvmopts.removeIf(option -> toRemove.contains(new JvmOption(option).option));
+            bag1.setJvmOptions(jvmopts);
+            int now = jvmopts.size();
+            if (removed) {
+                part.setMessage(lsm.getString("deleted.message", (orig-now)));
```
Comment (Pandrex247) [inline nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/commands/DeleteJvmOptions.java]: Space
`orig - now`

File: appserver/admingui/common/src/main/java/org/glassfish/admingui/common/handlers/InstanceHandler.java [hunk 2]
```diff
@@ -157,12 +178,13 @@ public static void saveJvmOptionValues(HandlerContext handlerCtx) {
...
-        payload.put("target", target);
-        for (Map oneRow : options) {
-            String str = (String) oneRow.get(PROPERTY_VALUE);
-            String str1 = UtilHandlers.escapePropertyValue(str);         //refer to GLASSFISH-19069
-            ArrayList kv = getKeyValuePair(str1);
+    private static void prepareJvmOptionPayload(Map<String, Object> payload, String target, List<Map<String, Object>> options) {
+        payload.put(TARGET, target);
+        for (Map<String, Object> oneRow : options) {
+            String jvmOptionUnescaped = new JvmOption((String)oneRow.get(JVM_OPTION),
+                    (String)oneRow.get(MIN_VERSION), (String)oneRow.get(MAX_VERSION)).toString();
```
Comment (Pandrex247) [inline appserver/admingui/common/src/main/java/org/glassfish/admingui/common/handlers/InstanceHandler.java]: Space between cast class and method please.
`(String) oneRow.get(MIN_VERSION)`

File: nucleus/admin/launcher/src/main/java/com/sun/enterprise/admin/launcher/GFLauncher.java [hunk 3]
```diff
@@ -818,7 +819,8 @@ private void setupProfilerAndJvmOptions(MiniXmlParser parser) throws MiniXmlPars
                 parser.getProfilerJvmOptions(),
                 parser.getProfilerSystemProperties());
 
-        List<String> rawJvmOptions = parser.getJvmOptions();
+        List<String> rawJvmOptions = parser.getJvmOptions().stream().filter(fullOption -> JDK.isCorrectJDK(fullOption.minVersion, fullOption.maxVersion))
```
Comment (Pandrex247) [inline nucleus/admin/launcher/src/main/java/com/sun/enterprise/admin/launcher/GFLauncher.java]: Could possibly do with some braces for readability? ¯\\_(ツ)_/¯

Comment (arjantijms) [reply nucleus/admin/launcher/src/main/java/com/sun/enterprise/admin/launcher/GFLauncher.java]: Personally I think builder style would help more here:

```java
parser.getJvmOptions()
      .stream()
      .filter(fullOption -> JDK.isCorrectJDK(fullOption.minVersion, fullOption.maxVersion))
```

And even better:

```java
parser.getJvmOptions()
      .stream()
      .filter(fullOption -> isCorrectJDK(fullOption.minVersion, fullOption.maxVersion))
```

Comment (Pandrex247) [reply nucleus/admin/launcher/src/main/java/com/sun/enterprise/admin/launcher/GFLauncher.java]: Static imports are evil, don't do it Lenny!

File: nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/xml/MiniXmlParser.java [hunk 4]
```diff
@@ -232,7 +232,72 @@ public String getInstanceLogFilename() {
...
         return logFilename;
     }
-    
+
+    public static class JvmOption {
+        public final String option;
+        public final Optional<JDK.Version> minVersion;
+        public final Optional<JDK.Version> maxVersion;
+
+        private static final Pattern PATTERN = Pattern.compile("^\\[(.*)\\|(.*)\\](.*)");
```
[the remark this answers is not in the document]
Comment (lprimak) [reply nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/xml/MiniXmlParser.java]: Fixed

File: nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/xml/MiniXmlParser.java [hunk 5]
```diff
@@ -232,7 +232,72 @@ public String getInstanceLogFilename() {
...
+            }
+            return true;
+        }
+
+        @Override
+        public String toString() {
+            if (!minVersion.isPresent() && !maxVersion.isPresent()) {
+                return option;
+            }
+            return String.format("[%s|%s]%s", minVersion.isPresent()? minVersion.get(): "", maxVersion.isPresent()? maxVersion.get(): "", option);
```
[the remark this answers is not in the document]
Comment (lprimak) [reply nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/xml/MiniXmlParser.java]: Not sure if I agree, but I really don't care that much.
```Question? AnswerA : AnswrB``` vs.
```Question ? AnserA : AnserB

I think Question mark belongs together with the question itself.
Makes sense?

File: nucleus/common/common-util/src/main/java/com/sun/enterprise/util/JDK.java [hunk 6]
```diff
@@ -98,36 +75,173 @@ public static int getUpdate() {
...
-        } else if (major == version.getMajor()) {
-            if (minor > version.getMinor()) {
+    public static class Version {
+        private final int major;
+        private final Optional<Integer> minor;
+        private final Optional<Integer> subminor;
+        private final Optional<Integer> update;
+
+        private Version(String string) {
+            String[] split = string.split("[\\._u\\-]+");
```
[the remark this answers is not in the document]
Comment (lprimak) [reply nucleus/common/common-util/src/main/java/com/sun/enterprise/util/JDK.java]: Fixed

File: nucleus/common/common-util/src/main/java/com/sun/enterprise/util/JDK.java [hunk 7]
```diff
@@ -98,36 +75,173 @@ public static int getUpdate() {
...
+                sb.append('.').append(update.get());
+            }
+            return sb.toString();
+        }
     }
 
-    public boolean newerOrEquals(JDK version) {
-        return newerThan(version) || equals(version);
+    public static Version getVersion(String string) {
+        if (string != null && string.matches("([0-9]+[\\._u\\-]+)*[0-9]+")) {
```
[the remark this answers is not in the document]
Comment (lprimak) [reply nucleus/common/common-util/src/main/java/com/sun/enterprise/util/JDK.java]: Fixed

File: nucleus/common/common-util/src/main/java/com/sun/enterprise/util/JDK.java [hunk 8]
```diff
@@ -98,36 +75,173 @@ public static int getUpdate() {
...
     }
 
-    public boolean olderOrEquals(JDK version) {
-        return !newerThan(version);
+    public static boolean isCorrectJDK(Optional<Version> minVersion, Optional<Version> maxVersion) {
+        boolean correctJDK = true;
+        if (minVersion.isPresent()) {
+            correctJDK = JDK_VERSION.newerOrEquals(minVersion.get());
+        }
+        if (correctJDK == true && maxVersion.isPresent()) {
```
[the remark this answers is not in the document]
Comment (lprimak) [reply nucleus/common/common-util/src/main/java/com/sun/enterprise/util/JDK.java]: Fixed

File: nucleus/admin/launcher/src/main/java/com/sun/enterprise/admin/launcher/GFLauncher.java (hunk 3 above)
Comment (lprimak) [reply nucleus/admin/launcher/src/main/java/com/sun/enterprise/admin/launcher/GFLauncher.java]: Fixed

**Comment by lprimak** [comment]
All fixed up

**Comment by lprimak** [comment]
jenkins test

File: nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/xml/MiniXmlParser.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Pandrex247) [reply nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/xml/MiniXmlParser.java]: Depends if you see the `?` as an operator or not.
I see it as an operator, in a similar vein to `+` and `=`.

[the remark this answers is not in the document]
Comment (Pandrex247) [reply nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/xml/MiniXmlParser.java]: We also use a space everywhere else we use ternary operators.

[the remark this answers is not in the document]
Comment (arjantijms) [reply nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/xml/MiniXmlParser.java]: I know this an ancient thread, but just to add to this; there's a Sun code convention for it. Not sure I like it so much, but I've learned to just accept it ;)

**Pull request merged**